---
title: Gut Microbiome Diversity Visualization
author: Morgan Essex
date: '2020-05-10'
slug: tidyverse-gut-diversity
output:
  blogdown::html_page:
    fig_caption: true
    toc: true
categories: 
    - tutorial
    - microbiome
tags:
    - ecrc
    - forslund-lab
    - ggplot
    - gut microbiome
    - programming
    - r-stats
    - tidyverse
    - tutorial
summary:
    A short workflow for generating alpha and beta diversity plots using taxonomic gut microbiome data from patients with autoimmune diseases.
---


<div id="TOC">
<ul>
<li><a href="#alpha-diversity-boxplots">Alpha Diversity Boxplots</a></li>
<li><a href="#beta-diversity-pcoa-plot">Beta Diversity PCoA Plot</a></li>
<li><a href="#summary">Summary</a></li>
</ul>
</div>

<p>Alpha and beta diversity are ecological metrics frequently part of exploratory microbiome analysis. The former measures the taxonomic variance <em>within</em> a sample, while the latter measures how samples vary <em>between</em> one another. There are different ways to measure both indices.</p>
<p>I’m going to use some actual (anonymized) 16S taxonomic data that I’m working with. It’s a composite dataset of patients (N = 222) presenting with any combination of <a href="https://en.wikipedia.org/wiki/Crohn%27s_disease">Crohn’s disease</a> (IBD), <a href="https://www.mayoclinic.org/diseases-conditions/ankylosing-spondylitis/symptoms-causes/syc-20354808">ankylosing spondylitis</a> (SpA), or acute anterior <a href="https://en.wikipedia.org/wiki/Uveitis#Types">uveitis</a> (AAU), pooled with a subset (N = 107) of healthy control samples from the <a href="https://academic.oup.com/cardiovascres/advance-article-abstract/doi/10.1093/cvr/cvaa128/5831292?redirectedFrom=fulltext">recently published</a> Collect study. We’ll be working with the genus-level rarefied data that has been transformed to relative abundances and filtered for prevalence in &gt; 5% of samples.</p>
<p>First, libraries. Like <code>ggsci</code> and <code>ggpubr</code>, <a href="https://github.com/slowkow/ggrepel">ggrepel</a> is another plotting package using <code>ggplot2</code> syntax that’s excellent for adding customizable labels to plots, which we’ll do later.</p>
<pre class="r"><code>library(tidyverse)
library(magrittr)
library(ggsci) 
library(ggrepel)
library(ggpubr)</code></pre>
<div id="alpha-diversity-boxplots" class="section level2">
<h2>Alpha Diversity Boxplots</h2>
<p>The abundance tables I’m going to work with were generated from raw 16S reads using <a href="https://microbiomejournal.biomedcentral.com/articles/10.1186/2049-2618-2-30">LoTuS</a> and rarefied using the <a href="https://academic.oup.com/bioinformatics/article/33/16/2594/3111845">rarefaction toolkit</a> (rtk), which calculates per-sample diversity metrics so we don’t have to do that, and I’ll just be reading in the anonymized metadata table instead.</p>
<p>There are packages for calculating diversity indices in R from an abundance table too if that’s needed, for example <code>phyloseq</code> or <code>microbiome</code>.</p>
<p>The richness is an OTU count (how many OTUs, here genera, in a sample), while the <a href="https://en.wiktionary.org/wiki/Shannon_entropy">Shannon entropy</a> is also informed by abundances and yields a measure of how evenly OTUs are distributed across a sample; a high Shannon entropy implies that some OTUs have similar abundance levels, while a low entropy indicates that some OTUs dominate over others. A lower Shannon diversity is frequently associated with chronic disease.</p>
<pre class="r"><code>meta.tbl &lt;- &#39;data/meta.tbl.rds&#39; %&gt;%
    readRDS() %&gt;%
    print()</code></pre>
<pre><code>## # A tibble: 329 x 4
##    Cohort Phenotype Shannon Richness
##    &lt;chr&gt;  &lt;chr&gt;       &lt;dbl&gt;    &lt;dbl&gt;
##  1 AAU    SpA+AAU      2.31       43
##  2 AAU    SpA+AAU      2.43       49
##  3 AAU    SpA+AAU      2.37       45
##  4 AAU    AAU          2.37       39
##  5 AAU    SpA+AAU      2.22       37
##  6 AAU    AAU          2.65       70
##  7 AAU    SpA+AAU      2.73       47
##  8 AAU    SpA+AAU      2.47       46
##  9 AAU    SpA+AAU      2.72       58
## 10 AAU    AAU          1.60       34
## # … with 319 more rows</code></pre>
<p>That’s actually all we need to generate a quick plot for ourselves.</p>
<pre class="r"><code>meta.tbl %&gt;%
    ggplot(aes(x = Cohort, 
               y = Shannon, 
               fill = Cohort)) +
    geom_boxplot(alpha = 0.5) +
    geom_jitter(alpha = 0.7, size = 2, width = 0.2) +
    labs(title = &#39;Alpha Diversity by Cohort&#39;, 
         y = &#39;Shannon Entropy&#39;, x = &#39;&#39;) +
    ggsci::scale_fill_futurama() +
    ggpubr::theme_pubr(legend = &#39;none&#39;)</code></pre>
<p><img src="/posts/tidyverse-gut-diversity/index_files/figure-html/alpha.cohort.1-1.png" width="672" /></p>
<p>I’ve opted to for <code>geom_boxplot()</code> with a <code>geom_point()</code> overlay to get a better idea of comparison sizes. This is a perfectly nice plot that already confirms an important hypothesis about this data: patients suffering from immune-mediated inflammatory diseases have lower microbial diversity than controls, with the proximity of the inflammation to the gut microbiota likely playing a role (hence why IBD samples are less diverse than SpA samples).</p>
<p>If we wanted to publish this plot or use it in a presentation, there are a few things we can do to make it even nicer, like making the labels more informative and adding statistics.</p>
<p>I’ll create a new <code>plot.data</code> object and use my favorite <code>mutate_at()</code> + the <a href="https://forcats.tidyverse.org/">forcats</a> package <strong>for</strong> dealing with <strong>cat</strong>egorical variables to do this. First I want to change the Cohort names, so I’ll use <code>fct_recode()</code> to supply the new names, and then <code>fct_relevel()</code> to order them how I want them to appear on the plot from left to right. Lastly, I’ll order the Phenotype names as well, since we’ll eventually plot those on the x-axis too.</p>
<pre class="r"><code>plot.data &lt;- meta.tbl %&gt;%
    # `new` = &#39;old&#39; ; note difference between `` and &#39;&#39; !
    mutate_at(vars(Cohort), ~ fct_recode(., 
                                         Collect = &#39;Collect&#39;,
                                         `Crohn&#39;s\nDisease` = &#39;IBD&#39;,
                                         Spondyloarthritis = &#39;SpA&#39;,
                                         Uveitis = &#39;AAU&#39;)) %&gt;%
    # takes strings, first var will be leftmost boxplot
    mutate_at(vars(Cohort), ~ fct_relevel(., 
                                          &#39;Collect&#39;,
                                          &#39;Spondyloarthritis&#39;,
                                          &quot;Crohn&#39;s\nDisease&quot;,
                                          &#39;Uveitis&#39;)) %&gt;%
    mutate_at(vars(Phenotype), ~ fct_relevel(., 
                                             &#39;Control&#39;,
                                             &#39;IBD&#39;,&#39;IBD+SpA&#39;,&#39;IBD+AAU&#39;,
                                             &#39;IBD+SpA+AAU&#39;,&#39;SpA&#39;,&#39;SpA+AAU&#39;,
                                             &#39;AAU&#39;)) %&gt;%
    print()</code></pre>
<pre><code>## # A tibble: 329 x 4
##    Cohort  Phenotype Shannon Richness
##    &lt;fct&gt;   &lt;fct&gt;       &lt;dbl&gt;    &lt;dbl&gt;
##  1 Uveitis SpA+AAU      2.31       43
##  2 Uveitis SpA+AAU      2.43       49
##  3 Uveitis SpA+AAU      2.37       45
##  4 Uveitis AAU          2.37       39
##  5 Uveitis SpA+AAU      2.22       37
##  6 Uveitis AAU          2.65       70
##  7 Uveitis SpA+AAU      2.73       47
##  8 Uveitis SpA+AAU      2.47       46
##  9 Uveitis SpA+AAU      2.72       58
## 10 Uveitis AAU          1.60       34
## # … with 319 more rows</code></pre>
<p>The next thing we want to do is add statistical significance between the cohorts, using the <a href="https://rpkgs.datanovia.com/ggpubr/index.html">ggpubr</a> package (the same one I’ll use for custom color palettes). It’s a good package with it’s own tweaked version of e.g. boxplots that’s potentially worth checking out.</p>
<p>For statistical significance, we’ll want to specify a list of comparisons to make, which we’ll give as the <code>comparisons</code> argument to the <code>stat_compare_means()</code> function. I’ve elected to print the stars, but replacing ..p.signif.. with ..p.val.. in <code>aes(label = ....)</code> will add the actual values. I’ve added a shape aesthetic <code>aes(shape = Cohort)</code> as well.</p>
<pre class="r"><code>to.compare &lt;- list(c(&#39;Collect&#39;,&#39;Spondyloarthritis&#39;),
                  c(&#39;Collect&#39;,&quot;Crohn&#39;s\nDisease&quot;),
                  c(&#39;Collect&#39;,&#39;Uveitis&#39;))

plot.data %&gt;%
    ggplot(aes(x = Cohort, 
               y = Shannon, 
               fill = Cohort)) +
    geom_boxplot(alpha = 0.5) +
    geom_jitter(aes(shape = Cohort), 
                alpha = 0.7, size = 2, width = 0.2) +
    labs(title = &#39;Alpha Diversity by Cohort&#39;, 
         y = &#39;Shannon Entropy&#39;, x = &#39;&#39;) +
    ggsci::scale_fill_futurama() +
    ggpubr::stat_compare_means(comparisons = to.compare, size = 4, 
                               tip.length = 0, step.increase = 0.15,
                               aes(label = ..p.signif..)) +
    ggpubr::theme_pubr(legend = &#39;none&#39;)</code></pre>
<p><img src="/posts/tidyverse-gut-diversity/index_files/figure-html/alpha.cohort.2-1.png" width="672" /></p>
<p>As I mentioned at the beginning, this is a composite dataset, and the Cohort variable refers to the <em>primary</em> disease a patient has been diagnosed with/the cohort they were recruited from. Let’s switch to looking at the alpha diversity across different <em>phenotypes</em> rather than cohorts, since many patients are suffering from two or more autoimmune diseases.</p>
<p>Here I’m going to use a lot more custom code that I can dive into after, which will hopefully illuminate <code>ggplot2</code> syntax a little bit or perhaps serve as a skeleton for your own boxplots.</p>
<pre class="r"><code>to.compare &lt;- list(c(&#39;Control&#39;,&#39;IBD&#39;), c(&#39;Control&#39;,&#39;IBD+SpA&#39;), c(&#39;Control&#39;,&#39;IBD+AAU&#39;),
                  c(&#39;Control&#39;,&#39;IBD+SpA+AAU&#39;), c(&#39;Control&#39;,&#39;SpA&#39;), c(&#39;Control&#39;,&#39;SpA+AAU&#39;), 
                  c(&#39;Control&#39;,&#39;AAU&#39;))

plot.data %&gt;%
    ggplot(aes(x = Phenotype, 
               y = Shannon, 
               fill = Phenotype)) +
    geom_boxplot(alpha = 0.5, outlier.alpha = 0) +
    geom_jitter(aes(shape = Cohort), 
                alpha = 0.5, size = 2, width = 0.2) +
    ggsci::scale_fill_d3() +
    labs(title = &#39;Alpha Diversity by Phenotype&#39;, 
         y = &#39;Shannon Entropy&#39;, x = &#39;&#39;) +
    # remove color/fill legend and change size of Cohort point legend
    guides(fill = &#39;none&#39;,
           shape = guide_legend(override.aes = list(size = 3))) +
    ggpubr::stat_compare_means(comparisons = to.compare, size = 3, 
                               tip.length = 0.03, step.increase = 0.15,
                               aes(label = ..p.signif..)) +
    theme_bw() +
    theme(legend.position = &#39;top&#39;,
          legend.justification = &#39;left&#39;,
          # remove major and minor grid lines
          panel.grid = element_blank(), 
          # remove bounding box around plot
          panel.border = element_blank(), 
          # add x+y axes back
          axis.line = element_line(color = &#39;black&#39;)) </code></pre>
<p><img src="/posts/tidyverse-gut-diversity/index_files/figure-html/alpha.facet-1.png" width="672" /></p>
<p>There’s not a ton of additional biological information in this plot and the statistical significance is probably overkill at this point, but you get the picture. It’s a nice way to embed additional information if needed.</p>
<p>Aesthetically, instead of using the <code>theme_pubr()</code> from <code>ggpubr</code>, I’ve opted for the built-in <code>theme_bw()</code> and customized it to look a lot like the pubr theme by using the <a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a> command - one really worth exploring the potential of on your own. I’ve also used it to put the legend horizontally at the top of the figure instead of off to the right. Using the <code>guides()</code> command, I’ve increased the size of the display points for ease of reading, and removed the color/fill legend which would just be superfluous since that corresponds to Phenotype which is already printed on the x-axis.</p>
<p>Lastly, what if we wanted to look at both of these variables in a slightly different way? Facetting is a really useful <code>ggplot2</code> feature that would allow us to do just that. For this plot, I want to filter out the controls so that we’re just looking at the autoimmune diseases, and I want to display Phenotype on the x-axis facetted by Cohort. Using the <code>scales = 'free_x'</code> parameter in the <code>facet_wrap()</code> command allows the x-axis labels to differ between facets, which they obviously will for this sort of analysis. I’ve tilted the x-axis text as well using the <code>axis.text.x</code> parameter of <code>theme()</code>.</p>
<pre class="r"><code>plot.data %&gt;%
    # remove controls from this visualization
    filter(Phenotype != &#39;Control&#39;) %&gt;%
    ggplot(aes(x = Phenotype, 
               y = Shannon, 
               fill = Phenotype)) +
    geom_boxplot(alpha = 0.5) +
    geom_jitter(aes(shape = Cohort), 
                alpha = 0.7, size = 2, width = 0.2) +
    ggsci::scale_fill_uchicago() +
    # group plots into facets by cohort, 
    facet_wrap(~ Cohort, scales = &#39;free_x&#39;) +
    labs(title = &#39;Alpha Diversity by Cohort and Phenotype&#39;, 
         y = &#39;Shannon Entropy&#39;, x = &#39;&#39;, shape = &#39;Cohort&#39;) +
    guides(fill = &#39;none&#39;,
           shape = guide_legend(override.aes = list(size = 3))) +
    theme_bw() +
    theme(legend.position = &#39;top&#39;,
          legend.justification = &#39;left&#39;,
          # tilt the x-axis text 
          axis.text.x = element_text(angle = -45, hjust = 0)) </code></pre>
<p><img src="/posts/tidyverse-gut-diversity/index_files/figure-html/alpha.phenotype-1.png" width="672" /></p>
<pre class="r"><code>          # tilt it the other way
          # axis.text.x = element_text(angle = 45, hjust = 1)) </code></pre>
</div>
<div id="beta-diversity-pcoa-plot" class="section level2">
<h2>Beta Diversity PCoA Plot</h2>
<p>The <a href="https://en.wikipedia.org/wiki/Beta_diversity">beta diversity</a> is a measurement of OTU diversity between different ecosystems, here different patient microbiota. The Bray-Curtis dissimilarity is a semimetric statistic bounded between 0 and 1 (0 = same OTUs in same abundances, 1 = samples do not share any OTUs) quantifying the pairwise compositional overlap between samples.</p>
<p>Once dissimilarities have been computed, a PCoA ordination creates an n-dimensional (usually n = 2) mapping of the dissimilarities between samples such that the calculated dissimilarity between samples is preserved and able to be visualized in a lower-dimensional space.</p>
<p>For this, I’ll read in the actual abundance table which has the Cohort variable as a label in the first column.</p>
<!-- Additionally, a **permutational multivariate analysis of variance (PERMANOVA)** significance test can be carried out on the distance matrix to assess whether the variance explained by a given predictor (e.g. the disease phenotype) is significant or not. -->
<!-- **NOTE:** the PERMANOVA test does not distinguish between diffferences among groups which are the result of location (a "true" difference in multivariate space) or dispersion (variance homogeneity), and significance values should be interpreted with caution. See the Assumptions section of the original method [linked here](https://onlinelibrary.wiley.com/doi/full/10.1111/j.1442-9993.2001.01070.pp.x). -->
<pre class="r"><code>taxa.tbl &lt;- &#39;data/taxa.tbl.rds&#39; %&gt;%
    readRDS() %&gt;%
    print()</code></pre>
<pre><code>## # A tibble: 329 x 134
##    Cohort Acidaminococcus Actinomyces Agathobacter Akkermansia Alistipes
##    &lt;chr&gt;            &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;     &lt;dbl&gt;
##  1 AAU             0                0            0    0.000432  0.000865
##  2 AAU             0                0            0    0.0182    0.00389 
##  3 AAU             0                0            0    0         0.0471  
##  4 AAU             0                0            0    0         0.000432
##  5 AAU             0                0            0    0         0       
##  6 AAU             0                0            0    0         0.0147  
##  7 AAU             0                0            0    0         0.0104  
##  8 AAU             0.0233           0            0    0         0.00173 
##  9 AAU             0                0            0    0.0519    0.00692 
## 10 AAU             0                0            0    0         0       
## # … with 319 more rows, and 128 more variables: Allisonella &lt;dbl&gt;,
## #   Allobacillus &lt;dbl&gt;, Alloprevotella &lt;dbl&gt;, Anaerofilum &lt;dbl&gt;,
## #   Anaerostipes &lt;dbl&gt;, Anaerotruncus &lt;dbl&gt;, Atopobium &lt;dbl&gt;,
## #   Bacteroides &lt;dbl&gt;, Barnesiella &lt;dbl&gt;, Bifidobacterium &lt;dbl&gt;,
## #   Bilophila &lt;dbl&gt;, Blautia &lt;dbl&gt;, Butyricicoccus &lt;dbl&gt;, Butyricimonas &lt;dbl&gt;,
## #   `CAG-352` &lt;dbl&gt;, `CAG-56` &lt;dbl&gt;, Campylobacter &lt;dbl&gt;,
## #   Candidatus_Soleaferrea &lt;dbl&gt;, Catenibacterium &lt;dbl&gt;, Collinsella &lt;dbl&gt;,
## #   Coprobacter &lt;dbl&gt;, `Defluviitaleaceae_UCG-011` &lt;dbl&gt;, Desulfovibrio &lt;dbl&gt;,
## #   Dialister &lt;dbl&gt;, Dorea &lt;dbl&gt;, DTU089 &lt;dbl&gt;, Eggerthella &lt;dbl&gt;,
## #   Eisenbergiella &lt;dbl&gt;, Enterococcus &lt;dbl&gt;, Enterorhabdus &lt;dbl&gt;,
## #   Erysipelatoclostridium &lt;dbl&gt;, `Erysipelotrichaceae_UCG-003` &lt;dbl&gt;,
## #   `Escherichia-Shigella` &lt;dbl&gt;, Faecalibacterium &lt;dbl&gt;, Faecalibaculum &lt;dbl&gt;,
## #   Faecalitalea &lt;dbl&gt;, `Family_XIII_UCG-001` &lt;dbl&gt;, Flavonifractor &lt;dbl&gt;,
## #   Fournierella &lt;dbl&gt;, Fusicatenibacter &lt;dbl&gt;, Fusobacterium &lt;dbl&gt;,
## #   `GCA-900066575` &lt;dbl&gt;, Gordonibacter &lt;dbl&gt;, Haemophilus &lt;dbl&gt;,
## #   Helicobacter &lt;dbl&gt;, Holdemanella &lt;dbl&gt;, Holdemania &lt;dbl&gt;, Howardella &lt;dbl&gt;,
## #   Hungatella &lt;dbl&gt;, Imtechella &lt;dbl&gt;, Intestinibacter &lt;dbl&gt;,
## #   Intestinimonas &lt;dbl&gt;, Lachnoclostridium &lt;dbl&gt;, Lachnospira &lt;dbl&gt;,
## #   `Lachnospiraceae_UCG-001` &lt;dbl&gt;, `Lachnospiraceae_UCG-003` &lt;dbl&gt;,
## #   `Lachnospiraceae_UCG-004` &lt;dbl&gt;, `Lachnospiraceae_UCG-008` &lt;dbl&gt;,
## #   `Lachnospiraceae_UCG-010` &lt;dbl&gt;, Lactobacillus &lt;dbl&gt;, Lactococcus &lt;dbl&gt;,
## #   Marvinbryantia &lt;dbl&gt;, Megasphaera &lt;dbl&gt;, Methanobrevibacter &lt;dbl&gt;,
## #   Mitsuokella &lt;dbl&gt;, Moryella &lt;dbl&gt;, Negativibacillus &lt;dbl&gt;,
## #   Odoribacter &lt;dbl&gt;, Olsenella &lt;dbl&gt;, Oscillibacter &lt;dbl&gt;,
## #   Oscillospira &lt;dbl&gt;, Papillibacter &lt;dbl&gt;, Parabacteroides &lt;dbl&gt;,
## #   Paraprevotella &lt;dbl&gt;, Parasutterella &lt;dbl&gt;, Peptococcus &lt;dbl&gt;,
## #   Phascolarctobacterium &lt;dbl&gt;, Phocea &lt;dbl&gt;, Prevotella &lt;dbl&gt;,
## #   `Prevotellaceae_UCG-001` &lt;dbl&gt;, Romboutsia &lt;dbl&gt;, Roseburia &lt;dbl&gt;,
## #   Ruminiclostridium &lt;dbl&gt;, `Ruminococcaceae_UCG-002` &lt;dbl&gt;,
## #   `Ruminococcaceae_UCG-003` &lt;dbl&gt;, `Ruminococcaceae_UCG-004` &lt;dbl&gt;,
## #   `Ruminococcaceae_UCG-005` &lt;dbl&gt;, `Ruminococcaceae_UCG-007` &lt;dbl&gt;,
## #   `Ruminococcaceae_UCG-010` &lt;dbl&gt;, `Ruminococcaceae_UCG-013` &lt;dbl&gt;,
## #   `Ruminococcaceae_UCG-014` &lt;dbl&gt;, Sellimonas &lt;dbl&gt;, Senegalimassilia &lt;dbl&gt;,
## #   Shuttleworthia &lt;dbl&gt;, Slackia &lt;dbl&gt;, Solobacterium &lt;dbl&gt;,
## #   Streptococcus &lt;dbl&gt;, Subdoligranulum &lt;dbl&gt;, Sutterella &lt;dbl&gt;,
## #   Terrisporobacter &lt;dbl&gt;, …</code></pre>
<p>We’ll use the <code>vegdist()</code> function from the <code>vegan</code> package to calculate the Bray-Curtis dissimilarites and save it in a <code>taxa.dist</code> object.</p>
<pre class="r"><code>taxa.dist &lt;- taxa.tbl %&gt;%
    select_if(is.double) %&gt;%
    vegan::vegdist(method = &#39;bray&#39;,
                   diag = TRUE, upper = TRUE)

head(taxa.dist)</code></pre>
<pre><code>## [1] 0.4824903 0.2733564 0.3601383 0.3484652 0.3290071 0.4795587</code></pre>
<p>From that, we can use the <code>cmdscale()</code> function from the <code>stats</code> package to obtain the point coordinates in PCoA space. We’ll add the Cohort labels back, since we’ll want to use them to distinguish points in our visualization.</p>
<pre class="r"><code>taxa.pcoa &lt;- taxa.dist %&gt;%
    stats::cmdscale(k = 2) %&gt;%
    as_tibble(names_repair = &#39;check_unique&#39;) %&gt;%
    add_column(Cohort = taxa.tbl$Cohort,
               .before = 1) %&gt;%
    print()</code></pre>
<pre><code>## # A tibble: 329 x 3
##    Cohort      V1       V2
##    &lt;chr&gt;    &lt;dbl&gt;    &lt;dbl&gt;
##  1 AAU    -0.0769  0.124  
##  2 AAU     0.173  -0.0781 
##  3 AAU    -0.216   0.0341 
##  4 AAU    -0.0196  0.0102 
##  5 AAU    -0.219   0.0795 
##  6 AAU    -0.204  -0.00588
##  7 AAU    -0.0754 -0.184  
##  8 AAU     0.0588  0.00364
##  9 AAU    -0.141  -0.0424 
## 10 AAU    -0.245   0.187  
## # … with 319 more rows</code></pre>
<p>Something useful I got in the habit of doing for my PCoA plots is adding the <strong>centroid</strong> of each group, i.e. the mean. I’m going to calculate those and save them in a separate <code>pcoa.centroids</code> variable. I’ve added a Pt.Type column that we’ll also use in our visualization to distinguish centroid points from the other data points (patient samples) in our plot.</p>
<pre class="r"><code>pcoa.centroids &lt;- taxa.pcoa %&gt;%
    group_by(Cohort) %&gt;%
    nest() %&gt;%
    ungroup() %&gt;%
    # calculate centroids for each Cohort
    mutate(V1 = map_dbl(data, ~ mean(.$V1))) %&gt;%
    mutate(V2 = map_dbl(data, ~ mean(.$V2))) %&gt;%
    # remove column of original points
    select(-data) %&gt;%
    add_column(Pt.Type = &#39;Group Mean&#39;) %&gt;%
    print()</code></pre>
<pre><code>## # A tibble: 4 x 4
##   Cohort       V1       V2 Pt.Type   
##   &lt;chr&gt;     &lt;dbl&gt;    &lt;dbl&gt; &lt;chr&gt;     
## 1 AAU     -0.0677 -0.0656  Group Mean
## 2 IBD     -0.194  -0.00146 Group Mean
## 3 SpA     -0.0978 -0.0424  Group Mean
## 4 Collect  0.224   0.0866  Group Mean</code></pre>
<p>There isn’t much to do to prepare this data for plotting, but we do want to add the centroids above to the <code>taxa.pcoa</code> values we got earlier. Additionally, since we added a Pt.Type variable to the centroids tibble, we’ll want to do the same for our individual coordinates before we can use the <code>bind_rows()</code> command. Just as with the alpha diversity plots, I’m using <code>fct_relevel()</code> to order the Cohort variable how we want it to appear in the final legend from top to bottom.</p>
<pre class="r"><code>plot.data &lt;- taxa.pcoa %&gt;%
    add_column(Pt.Type = &#39;Individual&#39;) %&gt;%
    bind_rows(pcoa.centroids) %&gt;%
    mutate_at(vars(Cohort), ~ fct_relevel(., &#39;SpA&#39;,&#39;IBD&#39;,&#39;AAU&#39;,&#39;Collect&#39;)) %&gt;%
    print()</code></pre>
<pre><code>## # A tibble: 333 x 4
##    Cohort      V1       V2 Pt.Type   
##    &lt;fct&gt;    &lt;dbl&gt;    &lt;dbl&gt; &lt;chr&gt;     
##  1 AAU    -0.0769  0.124   Individual
##  2 AAU     0.173  -0.0781  Individual
##  3 AAU    -0.216   0.0341  Individual
##  4 AAU    -0.0196  0.0102  Individual
##  5 AAU    -0.219   0.0795  Individual
##  6 AAU    -0.204  -0.00588 Individual
##  7 AAU    -0.0754 -0.184   Individual
##  8 AAU     0.0588  0.00364 Individual
##  9 AAU    -0.141  -0.0424  Individual
## 10 AAU    -0.245   0.187   Individual
## # … with 323 more rows</code></pre>
<p>Now our data is plot ready and we can use <code>geom_point()</code> again, this time to create a scatter plot of our PCoA coordinates. For visual appeal, I want to color these by cohort and use different shapes to represent the data points and the centroids, and make the centroids bigger. I also want to update the axis and legend titles with <code>labs()</code>, which I also used earlier, and finally use the <code>geom_label_repel()</code> function to label our centroids. This function takes a data frame or tibble as the first argument, and obtains the labels from a column in it as specified in the <code>mapping = aes(label = ...)</code> argument.</p>
<pre class="r"><code>plot.data %&gt;%
    ggplot(aes(x = V1, y = V2)) +
    geom_point(aes(shape = Pt.Type, size = Pt.Type,
                   color = Cohort, fill = Cohort)) +
    # 22 = filled square, 4 = x
    scale_shape_manual(values = c(22,4)) +
    # make centroids 2x larger
    scale_size_manual(values = c(4,2)) +
    ggsci::scale_color_futurama() +
    ggsci::scale_fill_futurama() +
    labs(title = &#39;Beta Diversity between Cohorts, genus-level&#39;,
         shape = &#39;Point Type&#39;, size = &#39;Point Type&#39;,
         x = &#39;PCoA Axis 1&#39;, y = &#39;PCoA Axis 2&#39;) + 
    ggrepel::geom_label_repel(dplyr::filter(plot.data, Pt.Type == &#39;Group Mean&#39;),
                              mapping = aes(label = Cohort), size = 9/.pt,
                              min.segment.length = unit(0, &#39;lines&#39;), force = 5) +
    theme_bw()</code></pre>
<p><img src="/posts/tidyverse-gut-diversity/index_files/figure-html/pcoa.1-1.png" width="672" /></p>
<p>Our beta diversity analysis confirms another biological hypothesis: patients with autoimmune diseases have more similar microbiota to one another than they do to healthy controls, which is why we see the largest separation on the x-axis and the Collect samples in their own little cloud.</p>
<p>Another thing I got used to doing on my PCoAs when possible is including a <code>stat_ellipse()</code> around the groupings in my PCoA. All we need to do is add two lines of code and our group separation becomes a lot clearer.</p>
<pre class="r"><code>plot.data %&gt;%
    ggplot(aes(x = V1, y = V2)) +
    # add ellipse around cohort groups
    stat_ellipse(geom = &#39;polygon&#39;, alpha = 0.2, 
                 aes(color = Cohort, fill = Cohort)) +
    geom_point(aes(shape = Pt.Type, size = Pt.Type,
                   color = Cohort, fill = Cohort)) +
    scale_shape_manual(values = c(22,4)) +
    scale_size_manual(values = c(4,2)) +
    ggsci::scale_color_futurama() +
    ggsci::scale_fill_futurama() +
    labs(title = &#39;Beta Diversity between Cohorts, genus-level&#39;,
         shape = &#39;Point Type&#39;, size = &#39;Point Type&#39;, 
         x = &#39;PCoA Axis 1&#39;, y = &#39;PCoA Axis 2&#39;) + 
    geom_label_repel(dplyr::filter(plot.data, Pt.Type == &#39;Group Mean&#39;),
                     mapping = aes(label = Cohort), size = 9/.pt,
                     min.segment.length = unit(0, &#39;lines&#39;), force = 5) +
    theme_bw()</code></pre>
<p><img src="/posts/tidyverse-gut-diversity/index_files/figure-html/pcoa.2-1.png" width="672" /></p>
</div>
<div id="summary" class="section level2">
<h2>Summary</h2>
<p>There are really two kinds of visual data analysis: that which is for your eyes only while you’re figuring stuff out, and that which is intended to communicate something once you’ve arrived at a conclusion. Though this was an exploratory analysis (for you), my actual first plots of this data were <em>much</em> simpler and uglier. However, the cool thing about <code>ggplot2</code>’s approach of using layered commands is that it can facilitate both the quick-and-dirty visualization <em>and</em> generate really elegant, customizable plots.</p>
<p>My motivation to get good at <code>ggplot2</code> grew out of a dislike of Illustrator/Inkscape work on a laptop. It’s the worst. I spent a lot of time playing around and reading the <a href="https://ggplot2.tidyverse.org/reference/index.html">documentation</a> website to get here (even before this I had only used the <code>theme()</code> option a few times without really understanding the capability), but I hope the code above serves as a more complete skeleton for your own exploration.</p>
<p>Mostly I hope you feel a little more comfortable in <code>ggplot2</code> now and can see the benefits of outsourcing more of the labor involved in data visualization to code.</p>
<style type="text/css">
img {
    display: block;
    margin-left: auto;
    margin-right: auto;
    width: 85%;
    padding: 10px
}
</style>
</div>
